Add the ability for Kodi to properly navigate Box Sets

the-black-eagle opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->
Kodi currently treats a box set as an album and lists all songs on that album in track and disc order.  This makes it difficult to navigate in sets that have many discs.
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This change adds the ability for Kodi to split an 'album' into it's separate discs and for each of those discs to be navigated independently.
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
Building and testing locally with multiple albums, including single disc albums, two and three disc sets, compilations and proper box sets.  Also tested with mis-tagged albums
<!--- Include details of your testing environment, and the tests you ran to -->
Ubuntu 18.04 64bit, ddd for debugging.  Multiple music sources.
<!--- see how your change affects other areas of the code, etc -->
It does not affect any other areas.
## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ X] All new and existing tests passed

the-black-eagle committed:
Initial work to add boxsets navigation

the-black-eagle committed:
Add listitems (DiscName and IsBoxset)

Remove unused code, switch to using pItem->SetArt

Add support for reading disc subtitles

Bump db version to 73. Update album & song tables for box sets

Remove strDiscSubtitle from artist table

DaveTBlake commented:
Marvellous,  a PR. :)
I'll have a look tomorrow, meanwhile if you could rebase without all the whitespace changes it would make me  happy person.

the-black-eagle commented:
I will try Dave.  I've already added support for TSST to CMusicInfoTagLoaderFFmpeg::Load and am about to address the db update.

the-black-eagle commented:
OK, so I'm stuck on fixing this whitespace issue!  I've rebased again, and git tells me that everything is up to date on both branches but it doesn't seem to have changed anything.

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -124,7 +124,7 @@ void CMusicDatabase::CreateTables()
               " strType text, strGender text, strDisambiguation text, "
               " strBorn text, strFormed text, strGenres text, strMoods text, "
-              " strStyles text, strInstruments text, strBiography text, "
+              " strDiscSubtitle text, strStyles text, strInstruments text, strBiography text, "
```
Discsubtitle in the artist table??? That is a boo boo for sure.

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -7486,9 +7491,9 @@ void CMusicDatabase::UpdateTables(int version)
-  SetMusicNeedsTagScan(60);
+  // The last schema change needing forced rescanning was 73.
+  // This is because Kodi can now read and process extra tags involved in the creation of box sets.
+  SetMusicNeedsTagScan(73);
```
OK force a rescan, otherwise no one will see anything.
Although one idea was during development to allow the rescanning to be optional when possible (even though to test the new features they need to rescan) and then bump this for say alpha and apply all the tag additions in one go.
There are going to be more tag additions (I have a list ready to do), so may decide to remove this for the Kodi PR depending on how near to PR I get with those additions.

DaveTBlake commented:
>OK, so I'm stuck on fixing this whitespace issue! I've rebased again, and git tells me that everything is up to date on both branches but it doesn't seem to have changed anything.

The whitespace is edited in your original commit, so you need to create that commit without those changes

DaveTBlake commented:
I am wondering what happens with `CTagLoaderTagLib::SetReleaseType` if the DISCSUBTITLE tag value is read after RELEASETYPE? I don't think there is any defined order for metadata. You need to gather all the tag data, and only then can you use them in combination. 

This could be done in `CMusicInfoScanner::FileItemsToAlbums` and/or creator of `CAlbum` (where artist credits are gathered from several tags). Note there are two distinct paths to how an album gets identified from tags depending on having mbid tags or not. This is historic and not ideal because of the inconsitencies it introduces (over say misssing ALBUMARTIST tags), and something I should look at one day. Anyway I'm just giving a pointer where to look.

Also no tag reading doesn't just handle different separators. See how they are done in `CAlbum::SetArtistCredits` using `StringUtils::SplitMulti`

the-black-eagle commented:
I don't think it matters in this instance which order the tags come in.  My code should handle it in either case.  If the disc subtitle has not yet been read, strDiscSubtitle will be empty, so it will get set to 'dummy' because of the 'boxset' in the tag.  If the disc subtitle is read at a later point, 'dummy' will be overwritten by the correct title.  If it's already been read then 'dummy' will be set only if it's empty.  Either path leads to a value in strDiscSubtitle.  _IF_ the actual disc title is empty and it is read later (which will overwrite 'dummy') it doesn't matter because the boxset flag is true and it will just get set back to dummy again.

the-black-eagle committed:
Show the disc subtitle in songview and MusicVisualisation

DaveTBlake commented on xbmc/music/infoscanner/MusicInfoScanner.cpp:
```diff
@@ -919,6 +923,58 @@ int CMusicInfoScanner::RetrieveMusicInfo(const std::string& strDirectory, CFileI
+        }
+      }
+      if (count < 3 && !album.bBoxedSet)  // must have more than 2 titled discs to be considered a boxset unless tag was set
+        album.bBoxedSet = false;    }
```
This only reached when !album.bBoxedSet anyway, setting does nothing?

the-black-eagle replied:
Yes it does, because bBoxedSet could have been forced to true by setting the tag, in which case, we don't want to set it to false if there are less than 3 discs.  If you take it out, you cannot make a twin disc set into a boxset.  Not sure why you would want to, but someone might.  It also messes up making boxsets from albums where the discs are in separate directories.

DaveTBlake commented on xbmc/music/infoscanner/MusicInfoScanner.cpp:
```diff
@@ -919,6 +923,58 @@ int CMusicInfoScanner::RetrieveMusicInfo(const std::string& strDirectory, CFileI
+  {
+    if (album.bCompilation  && !album.bBoxedSet)
+    {
+      album.bBoxedSet = false;
```
This only reached when !album.bBoxedSet anyway, setting does nothing?

the-black-eagle replied:
Yeah, setting the bool to false again is pointless, it should just be continue. I'll take line 943 out.  The check is important though, because without it, Kodi will start to create a boxset for a compilation (and it will show in the first node) but there won't be any discs associated with it.

DaveTBlake commented:
Fair enough regarding tag order processing I guess I just didn't expect it to be in the tag processing. I do need to look more closely overall, I'm just chipping a little bits at the moment (dsitracted with other things).

But I still need you to banish those cosmetic whitespace changes that make it look like you have involved areas of cocde that are nothing to do with this PR. Did the Estuary changes really come down to so little after all, or is another commit going to appear?

DaveTBlake commented on xbmc/music/tags/MusicInfoTag.cpp:
```diff
@@ -16,6 +16,7 @@
 #include "utils/Archive.h"
 #include "utils/StringUtils.h"
 #include "utils/Variant.h"
+#include "utils/log.h"
```
Needed?

the-black-eagle replied:
Nope.  And yeah, the Estuary changes were only small although the initial commit included some other files that I had edited whilst trying to get it to display what I wanted.

the-black-eagle committed:
Start to add JSON support

DaveTBlake commented:
Oh dear @the-black-eagle , I have a day off and when I get back the diff is even less readable. I think the problem is that you don't have a static clone of master branch in your repo or you rebased that to the current master but didn't rebase this branch. Either way it needs sorting out please.

DaveTBlake commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -2389,7 +2393,7 @@ msgctxt "#519"
 msgstr ""
 
-#empty string with id 520
+#: empty string 520
```
Revert to original text

DaveTBlake commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -12757,7 +12768,8 @@ msgctxt "#20340"
 msgstr ""
 
-#empty string with id 20341
+#: empty string #20341
```
Revert to original text/spaces

DaveTBlake commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -2192,7 +2192,11 @@ msgctxt "#470"
 
-#empty strings from id 471 to 473
+msgctxt "#471"
+msgid "Disc Title"
```
While it seems obvious to put this here near "disc" the 3-digit message text ids are old and impossible to search for, hence no new messages are to be added in this range.  Move to a higher ID

the-black-eagle commented:
I rebased both my master branch and this one to the current upstream master.  If I close this and open a new PR against my current master, will that work ?

DaveTBlake commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -2004,7 +2004,7 @@ msgstr ""
 #: addons/skin.estuary/xml/Variables.xml
 msgctxt "#427"
-msgid "Disc"
+msgid "Disc Number"
```
Are you sure #427 isn't used somewhere else where  adding " number" is incorrect? Need to take great care with the old 3-digit ID messages

DaveTBlake commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -21138,8 +21150,12 @@ msgctxt "#38073"
-#empty strings from id 38074 to 38099
-#strings 38074 to 38099 reserved for music library
+msgctxt "#38074"
+msgid "Box-sets"
```
"Box-sets", "Box sets".... hyphen consistency?

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -86,14 +86,14 @@ typedef struct
 /// Skins can use infolabels with <b>$INFO[infolabel]</b> or the <b>\<info\></b> tag. Scripts
 /// can read infolabels with <b>xbmc.getInfoLabel('infolabel')</b>.
-/// 
+///
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -86,14 +86,14 @@ typedef struct
 /// @todo [docs] Use links instead of bold values for infolabels/bools
 /// so we can use a link to point users when providing help
-/// 
+///
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -139,7 +139,7 @@ typedef struct
 ///     @return **True** if the info is empty.
 ///     @param info - infolabel
-///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \, 
+///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \,
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -153,7 +153,7 @@ typedef struct
 ///     @param info - infolabel
 ///     @param string - comparison string
-///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \, 
+///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \,
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -167,7 +167,7 @@ typedef struct
 ///     @param info - infolabel
 ///     @param substring - substring to check
-///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \, 
+///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \,
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -181,7 +181,7 @@ typedef struct
 ///     @param info - infolabel
 ///     @param substring - substring to check
-///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \, 
+///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \,
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -195,7 +195,7 @@ typedef struct
 ///     @param info - infolabel
 ///     @param substring - substring to check
-///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \, 
+///     @note **Example of info:** \link ListItem_Title `ListItem.Title` \endlink \,
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -550,7 +550,7 @@ const infomap integer_bools[] =  {{ "isequal",          INTEGER_IS_EQUAL },
 ///                  \anchor Player_FilenameAndPath
 ///                  _string_,
-///     @return The full path with filename of the currently 
+///     @return The full path with filename of the currently
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -583,14 +583,14 @@ const infomap integer_bools[] =  {{ "isequal",          INTEGER_IS_EQUAL },
 ///                  \anchor Player_TempoEnabled
 ///                  _boolean_,
-///     @return **True** if player supports tempo (i.e. speed up/down normal 
+///     @return **True** if player supports tempo (i.e. speed up/down normal
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -608,17 +608,17 @@ const infomap integer_bools[] =  {{ "isequal",          INTEGER_IS_EQUAL },
 ///                  \anchor Player_PlaySpeed
 ///                  _string_,
-///     @return The player playback speed with the format `%1.2f` (1.00 means normal 
+///     @return The player playback speed with the format `%1.2f` (1.00 means normal
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -608,17 +608,17 @@ const infomap integer_bools[] =  {{ "isequal",          INTEGER_IS_EQUAL },
+///     @return The player playback speed with the format `%1.2f` (1.00 means normal
 ///     playback speed).
-///     @note For Tempo\, the default range is 0.80 - 1.50 (it can be changed 
+///     @note For Tempo\, the default range is 0.80 - 1.50 (it can be changed
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -608,17 +608,17 @@ const infomap integer_bools[] =  {{ "isequal",          INTEGER_IS_EQUAL },
 ///                  \anchor Player_HasResolutions
 ///                  _boolean_,
-///     @return **True** if the player is allowed to switch resolution and refresh rate 
+///     @return **True** if the player is allowed to switch resolution and refresh rate
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -627,7 +627,7 @@ const infomap integer_bools[] =  {{ "isequal",          INTEGER_IS_EQUAL },
 ///                  \anchor Player_HasPrograms
 ///                  _boolean_,
-///     @return **True** if the media file being played has programs\, i.e. groups of streams. 
+///     @return **True** if the media file being played has programs\, i.e. groups of streams.
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -648,7 +648,7 @@ const infomap integer_bools[] =  {{ "isequal",          INTEGER_IS_EQUAL },
 ///     <p><hr>
 ///     @skinning_v18 **[New Infolabel]** \link Player_Icon `Player.Icon`\endlink
-///     <p>  
+///     <p>
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -1115,7 +1115,7 @@ const infomap weather[] =        {{ "isfetched",        WEATHER_IS_FETCHED },
 ///     @return **True** if PVR is supported from Kodi.
 ///     @note normally always true
-///     
+///
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -1575,7 +1575,7 @@ const infomap weather[] =        {{ "isfetched",        WEATHER_IS_FETCHED },
 ///                  \anchor System_GetBool
 ///                  _string_,
-///     @return The value of any standard system boolean setting. 
+///     @return The value of any standard system boolean setting.
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2010,9 +2010,9 @@ const infomap musicpartymode[] = {{ "enabled",           MUSICPM_ENABLED },
 ///     @return The name of the dj who remixed the selected song.
 ///     @todo So maybe rather than a row each have one entry for Role.XXXXX with composer\, arranger etc. as listed values
-///     @note MusicPlayer.Property(Role.any_custom_role) also works\, 
+///     @note MusicPlayer.Property(Role.any_custom_role) also works\,
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2010,9 +2010,9 @@ const infomap musicpartymode[] = {{ "enabled",           MUSICPM_ENABLED },
+///     @note MusicPlayer.Property(Role.any_custom_role) also works\,
 ///     where any_custom_role could be an instrument violin or some other production activity e.g. sound engineer.
-///     The roles listed (composer\, arranger etc.) are standard ones but there are many possible. 
+///     The roles listed (composer\, arranger etc.) are standard ones but there are many possible.
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2076,7 +2076,7 @@ const infomap musicpartymode[] = {{ "enabled",           MUSICPM_ENABLED },
 ///     @return Artist(s) of the song which has an offset `number` with respect
 ///     to the start of the playlist.
-///     @param number - the offset of the song with respect to 
+///     @param number - the offset of the song with respect to
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2107,7 +2107,7 @@ const infomap musicpartymode[] = {{ "enabled",           MUSICPM_ENABLED },
 ///     <p><hr>
 ///     @skinning_v18 **[New Infolabel]** \link MusicPlayer_Property_Artist_Type `MusicPlayer.Property(Artist_Type)`\endlink
-///     <p>  
+///     <p>
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2116,7 +2116,7 @@ const infomap musicpartymode[] = {{ "enabled",           MUSICPM_ENABLED },
 ///     <p><hr>
 ///     @skinning_v18 **[New Infolabel]** \link MusicPlayer_Property_Artist_Gender `MusicPlayer.Property(Artist_Gender)`\endlink
-///     <p>  
+///     <p>
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2125,7 +2125,7 @@ const infomap musicpartymode[] = {{ "enabled",           MUSICPM_ENABLED },
 ///     <p><hr>
 ///     @skinning_v18 **[New Infolabel]** \link MusicPlayer_Property_Artist_Disambiguation `MusicPlayer.Property(Artist_Disambiguation)`\endlink
-///     <p>  
+///     <p>
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2275,7 +2275,7 @@ const infomap musicpartymode[] = {{ "enabled",           MUSICPM_ENABLED },
 ///     <p><hr>
 ///     @skinning_v17 **[New Infolabel]** \link MusicPlayer_UserRating `MusicPlayer.UserRating`\endlink
-///     <p>  
+///     <p>
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2414,7 +2414,7 @@ const infomap musicpartymode[] = {{ "enabled",           MUSICPM_ENABLED },
 ///     @return The track number of the song with an offset `number`
 ///     with respect to start of the playlist.
-///     @param number - The offset number of the song with respect 
+///     @param number - The offset number of the song with respect
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3126,7 +3128,7 @@ const infomap videoplayer[] =    {{ "title",            VIDEOPLAYER_TITLE },
 ///     <p><hr>
 ///     @skinning_v18 **[New Infolabel]** \link RetroPlayer_StretchMode `RetroPlayer.StretchMode`\endlink
-///     <p>  
+///     <p>
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -2540,6 +2540,8 @@ const infomap musicplayer[] =    {{ "title",            MUSICPLAYER_TITLE },
                                   { "samplerate",       MUSICPLAYER_SAMPLERATE },
                                   { "codec",            MUSICPLAYER_CODEC },
                                   { "discnumber",       MUSICPLAYER_DISC_NUMBER },
+                                  { "discname",         MUSICPLAYER_DISC_NAME },
```
"disctitle" seems better than "discname" - people have names, songs  and discs have titles, and the data comes from DISCSUBTITLE tag. Going to live with the info label a long time so better to be happy with naming from begining.

DaveTBlake replied:
Also need to add comment to document addition
e.g. @skining-v19 ** [New Infolabel]**

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3140,7 +3142,7 @@ const infomap videoplayer[] =    {{ "title",            VIDEOPLAYER_TITLE },
 ///     <p><hr>
 ///     @skinning_v18 **[New Infolabel]** \link RetroPlayer_VideoRotation `RetroPlayer.VideoRotation`\endlink
-///     <p>  
+///     <p>
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3339,14 +3341,14 @@ const infomap mediacontainer[] = {{ "hasfiles",         CONTAINER_HASFILES },
 ///                  \anchor Container_NumItems
 ///                  _integer_,
-///     @return The number of items in the container or grouplist with given id excluding parent folder item. 
+///     @return The number of items in the container or grouplist with given id excluding parent folder item.
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3339,14 +3341,14 @@ const infomap mediacontainer[] = {{ "hasfiles",         CONTAINER_HASFILES },
 ///                  \anchor Container_NumAllItems
 ///                  _integer_,
-///     @return The number of all items in the container or grouplist with given id including parent folder item. 
+///     @return The number of all items in the container or grouplist with given id including parent folder item.
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3356,7 +3358,7 @@ const infomap mediacontainer[] = {{ "hasfiles",         CONTAINER_HASFILES },
 ///                  _integer_,
 ///     @return The Number of items in the container or grouplist with given id excluding all folder items.
-///     @note **Example:** pvr recordings folders\, parent ".." folder). 
+///     @note **Example:** pvr recordings folders\, parent ".." folder).
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3375,7 +3377,7 @@ const infomap mediacontainer[] = {{ "hasfiles",         CONTAINER_HASFILES },
 ///     with id (or current container if id is omitted).
 ///     @note This is slightly delayed from the actual scroll start. Use
-///     \ref Container_OnScrollNext "Container(id).OnScrollNext" or 
+///     \ref Container_OnScrollNext "Container(id).OnScrollNext" or
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3471,7 +3473,7 @@ const infomap container_bools[] ={{ "onnext",           CONTAINER_MOVE_NEXT },
 ///                  \anchor Container_CurrentItem
 ///                  _integer_,
-///     @return The current item in the container or grouplist with given id. 
+///     @return The current item in the container or grouplist with given id.
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3535,8 +3537,8 @@ const infomap container_ints[] = {{ "row",              CONTAINER_ROW },
 ///     @param offset - The offset for the listitem.
-///     @note That means if the last item of a list is focused\, `ListItemNoWrap(1)` 
-///     will be empty while `ListItem(1)` will return the first item of the list. 
+///     @note That means if the last item of a list is focused\, `ListItemNoWrap(1)`
```
Revert cosmetic whitespace changes

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3563,15 +3565,15 @@ const infomap container_ints[] = {{ "row",              CONTAINER_ROW },
 ///     @return **True** if the current container you are in contains the following:
-///       - <b>files</b> 
-///       - <b>songs</b> 
+///       - <b>files</b>
```
Revert cosmetic whitespace change

DaveTBlake commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3563,15 +3565,15 @@ const infomap container_ints[] = {{ "row",              CONTAINER_ROW },
+///       - <b>songs</b>
 ///       - <b>artists</b>
-///       - <b>albums</b> 
+///       - <b>albums</b>
```
Revert cosmetic whitespace change....
You can see all of them. Fix it!!

the-black-eagle replied:
All fixed......I think.

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -785,7 +785,8 @@
       { "name": "includesingles", "type": "boolean", "default": false },
-      { "name": "allroles", "type": "boolean", "default":false, "description": "Whether or not to include all roles when filtering by artist, rather than the default of excluding other contributions. When true it overrides any role filter value." }
+      { "name": "allroles", "type": "boolean", "default":false, "description": "Whether or not to include all roles when filtering by artist, rather than the default of excluding other contributions. When true it overrides any role filter value." },
+      { "name": "boxset", "type": "boolean", "default": false, "description": "When true only returns albums that are also boxsets"}
```
I know you are just beginning to look at this. 
I don't think boxset should be a separate parameter like this, but part of the general filtering instead (working the way that smart playlists work).

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -505,7 +505,7 @@
       "enum": [ "title", "description", "artist", "genre",
                 "theme", "mood", "style", "type", "albumlabel",
-                "rating", "votes", "userrating","year", "musicbrainzalbumid",
+                "rating", "votes", "userrating","year", "boxset", "musicbrainzalbumid",
```
New fields must be added at the end, in case any existing API consumers are using position of fields rather than name.

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -581,6 +581,7 @@
       "title": { "type": "string" },
       "artist": { "$ref": "Array.String" },
       "year": { "type": "integer" },
+      "boxset" : {"type": "integer" },
```
New fields at end

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -1503,7 +1506,7 @@
                 "episodeguide", "uniqueid", "dateadded", "channel", "channeltype", "hidden",
                 "locked", "channelnumber", "starttime", "endtime", "specialsortseason",
-                "specialsortepisode", "compilation", "releasetype", "albumreleasetype",
+                "specialsortepisode", "compilation", "boxset", "releasetype", "albumreleasetype",
```
At end....

DaveTBlake reviewed (commented):
I know just starting to look at JSON changes, but some early feedback.
Will want related stuff JSON in a separate commit, and skin changes separately too but most of the other commits can be squashed.  In fact it may help to have them squashed now, and certainly get the master they are compared against and the master you started from to match.

DaveTBlake commented:
>I rebased both my master branch and this one to the current upstream master. If I close this and open a new PR against my current master, will that work ?

That will just sever all the comments already made, so don't do that just yet.
If the masters have been updated then it looks to me like you have not rebased your ADD_BOXSETS_ABILITY branch against them. 
Rebase your local master, push that to your repo
git checkout master
git fetch upstream
git rebase upstream/master
git push -f origin master

Then rebase your local ADD_BOXSETS_ABILITY *against local master*, do some squashing while you are doing that. Finaly force push that to your repo, and this PR will update magically

DaveTBlake commented:
Sorry about the whitespace comment rampage, but it was begining to get to me!

DaveTBlake commented on xbmc/filesystem/MusicDatabaseDirectory/QueryParams.h:
```diff
@@ -23,7 +23,7 @@ namespace XFILE
       long GetSongId() { return m_idSong; }
       long GetYear() { return m_year; }
-
+      long GetDisc() { return m_idAlbum; }
```
m_disc not m_idAlbum?

DaveTBlake commented on xbmc/music/Album.h:
```diff
@@ -55,6 +55,7 @@ class CAlbum
     iUserrating = -1;
     iVotes = -1;
     iYear = -1;
+    bBoxedSet  = false;
```
Since everything else is "box set" I guess `bBoxSet` would be more consistent. 
Then again the field in the db is boxedset.....

DaveTBlake commented:
Good you have sorted out the master.diff problem. :)
Unfortuneately I'm unlikely to be able to really dig into this until the weekend (although you may get some passing comments if I get time). I want to really think about it, and I just haven't had chance.

Meanwhile MusicDatabase.cpp and MusicInfoScanner.cpp still have distracting whitespace changes. They will need to go before a real PR  (otherwise having cosmetic changes dotted all over a PR that adds a feature makes diffs and subsequent Git blame impossible to use). Yes MusicDatabase.cpp is in need of a big cosmetic clean-up of whitespace and out-of-date style (not to mention Doxygen comment documenting), but it needs to be a PR just for that. Anyway they need to go eventually, and it would make my life a whole lot easier if they went now.

the-black-eagle commented:
Sure.  There is no rush Dave, and by the weekend I should have cleaned up the whitespace issues, fixed a couple of silly errors I made today when tidying things and squashed the commits for that into one or two.  I'm well aware that a real PR will have to be spot-on to even have a chance of being accepted.

DaveTBlake commented on xbmc/interfaces/json-rpc/AudioLibrary.cpp:
```diff
@@ -220,9 +223,8 @@ JSONRPC_STATUS CAudioLibrary::GetAlbums(const std::string &method, ITransportLay
   {
-    for (CVariant::const_iterator_array field = parameterObject["properties"].begin_array(); 
-      field != parameterObject["properties"].end_array(); field++)
-      fields.insert(field->asString());
```
Restore missing insert

the-black-eagle committed:
Ensure strDiscSubtitle is not empty

the-black-eagle committed:
Only show boxset node if we have sets.  Add docs

DaveTBlake commented:
I have been doing some digging into Kodi dev history and here's the thing....
There is the wish to move away from folder type nodes to filter type nodes, work started but was never completed (usual thing), and music library particularly was never looked at with care.  Originally oblivious to this "design intent",  I have made use of folder type nodes in my work, (behaviour is slightly different to filter type, and I wanted that) but avoided adding any more decendant classes of CDirectoryNode.  Now you have added `CDirectoryNodeBoxsets`, `CDirectoryNodeBoxsetDiscs` and `CDirectoryNodeBoxsetDiscSongs`, and it is understandable why you took that approach looking at what currently exists **but**  really it is moving in the opposite direction to the one that is "wished" for. Not a show stopper, and I'm not going to block this on those grounds, but perhaps we can take this opportunity to work together to make the next leap more in the direction of the "wish". Big ask maybe, but hey why not? That is the only way Kodi improves, let's run with it a bit and see what we can do :)

The "dream" is of a more flexible node system that only needs directory nodes (if at all) for the base node types ("artists", "albums", "discs"(the new thing),  "songs"), then the user can build all other nodes using a rule based filter approach.  I think that means we do need CDirectoryNodeBoxsetDiscs, but that CDirectoryNodeBoxsets is a variety of CDirectoryNodeAlbum (and so are DirectoryNodeAlbumCompilations, DirectoryNodeAlbumRecentlyAdded etc. ) and DirectoryNodeBoxsetDiscSongs a variety of DirectoryNodeSong.

Being able to make _almost_ the same nodes using 
a) a CDirectoryNode descendant folder node (musicdb://boxedset), 
b) a folder node based on path with musicdb://albums/?boxedset=true [what approach (a) creates]
c) a filter node  type = "albums", rule = "boxedset is true"
is a little bonkers, and definitely confusing.

Have a poke about and see what you can learn.
BTW you can scream at me for making this harder, but I hope you feel like rising to the challenge (and benefit from avoiding a repeat of my original short-sightedness), or at least having a look. Failing that let me have a play and I can hopefully give some pointers. I have ideas already but need to check things out.

At the back of my mind is looking to gain the facility to optionally split albums/combine songs for all multi-disc releases, not just box sets (which would still stand as a much needed feature). That is, on those 2 CD albums be able click to view (and play) as all songs in a lump or separated as discs.

Also the facility to have navigation _optionally_ go genres > albums rather than fixed in stone as genres > artists, something I know some people want.

DaveTBlake commented:
The direction I'm looking entails making `CDirectoryNodeAlbum::GetChildType()` examine url options (or something) and switch child node type accordingly. 

For navigation from filter type nodes this connects to a change I am already working on - getting filter type nodes to pass their rules to the next level of item i.e. an "artists" smartplaylist/custom node showing artists with songs of genre X, when clicked shows just those albums with songs of genre X by that artist. At the moment you get all the albums by that artist, unlike a folder type node where the genre filter is applied all the way down. But not only do we need to pass the filter criteria, it also needs something to say navigate to  discs rather than songs. Still considering that, could use the "group by" xml tag value maybe... or it is a matter of transfering some rules to options at the end of the path.

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -8888,6 +8942,184 @@ bool CMusicDatabase::GetGenresJSON(CFileItemList& items, bool bSources)
+  // then split the disc names from strDiscTitles, count how many there are
+  // and loop around that many times to create a view with 'i' number of boxes,
+  // titled by the subtitles
+
```
`CMusicDatabase::GetBoxsetDiscs()` note to myself (because it may not be clear what I mean)
Check album first, then get disc title and number - both held in item along with repeated album data. Add both idAlbum and disc number in path to later use as index to get the songs for that disc in GetFilter

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -1923,6 +1938,35 @@ bool CMusicDatabase::GetArtistsBySong(int idSong, std::vector<int> &artists)
 }
 
+bool CMusicDatabase::GetSongByDiscSubtitleAndAlbum(const std::string& strDiscSubtitle, int idAlbum)
+{
```
A redundant function

the-black-eagle replied:
Yeah I know.  It was added originally when I wasn't sure exactly what routines I would need.

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -8888,6 +8942,184 @@ bool CMusicDatabase::GetGenresJSON(CFileItemList& items, bool bSources)
+}
+
+bool CMusicDatabase::GetBoxsetDiscSongs(const std::string& strBaseDir, CFileItemList& items)
+{
```
No need for this new method, with its bespoke path parsing and only fetching song data and not the artist credits etc. 
Fetch the song node data using GetSongsByWhere with the correct where clause being built from parameters/options by `GetFilter`.

DaveTBlake reviewed (commented):
I did warn that I don't nit-pick code style, as much as ravage design :)
I will be back with hints for how this could be done, so don't despair.

the-black-eagle commented:
> I have been doing some digging into Kodi dev history and here's the thing....
> There is the wish to move away from folder type nodes to filter type nodes, work started but was never completed (usual thing), and music library particularly was never looked at with care. Originally oblivious to this "design intent", I have made use of folder type nodes in my work, (behaviour is slightly different to filter type, and I wanted that) but avoided adding any more decendant classes of CDirectoryNode. Now you have added `CDirectoryNodeBoxsets`, `CDirectoryNodeBoxsetDiscs` and `CDirectoryNodeBoxsetDiscSongs`, and it is understandable why you took that approach looking at what currently exists **but** really it is moving in the opposite direction to the one that is "wished" for. Not a show stopper, and I'm not going to block this on those grounds, but perhaps we can take this opportunity to work together to make the next leap more in the direction of the "wish". Big ask maybe, but hey why not? That is the only way Kodi improves, let's run with it a bit and see what we can do :)
> 
> The "dream" is of a more flexible node system that only needs directory nodes (if at all) for the base node types ("artists", "albums", "discs"(the new thing), "songs"), then the user can build all other nodes using a rule based filter approach. I think that means we do need CDirectoryNodeBoxsetDiscs, but that CDirectoryNodeBoxsets is a variety of CDirectoryNodeAlbum (and so are DirectoryNodeAlbumCompilations, DirectoryNodeAlbumRecentlyAdded etc. ) and DirectoryNodeBoxsetDiscSongs a variety of DirectoryNodeSong.
> 
> Being able to make _almost_ the same nodes using
> a) a CDirectoryNode descendant folder node (musicdb://boxedset),
> b) a folder node based on path with musicdb://albums/?boxedset=true [what approach (a) creates]
> c) a filter node type = "albums", rule = "boxedset is true"
> is a little bonkers, and definitely confusing.
> 
> Have a poke about and see what you can learn.
> BTW you can scream at me for making this harder, but I hope you feel like rising to the challenge (and benefit from avoiding a repeat of my original short-sightedness), or at least having a look. Failing that let me have a play and I can hopefully give some pointers. I have ideas already but need to check things out.
> 
> At the back of my mind is looking to gain the facility to optionally split albums/combine songs for all multi-disc releases, not just box sets (which would still stand as a much needed feature). That is, on those 2 CD albums be able click to view (and play) as all songs in a lump or separated as discs.
> 
> Also the facility to have navigation _optionally_ go genres > albums rather than fixed in stone as genres > artists, something I know some people want.

The thing about filters and part of the reason I wasn't sure that a filter would be the correct way to go, is that as I understand it, a filter only applies at the current view 'level'.  I could probably make it work using a filter, but I'll have to look in much more detail at exactly how they work first.  That isn't a bad Idea anyway as you want the JSON implementation to do it by filtering so I know it's an area I need to be looking into.  
If I create a filter node for artists based on a path of 'StageMusicals' that shows me all the artists on those albums, but clicking on an artist shows me ALL the albums they are on, not just in that path. For that to work, you would have to apply the original filter again.  I'm just thinking out loud here but I think that would work as you would get artists filtered by path, then albums containing that artist, also filtered by path, then songs, yet again filtered by path as you could have the same song on another album in a different path.

The folder type node seemed to be the way to go by looking at how the videodatabase is set up.  A boxset is similar in construction to a TV show (Show -> seasons -> epsiodes) which seemed to correspond well with album -> discs -> songs so that is the main reason it is how it is.  I do agree though that the boxset node is basically a filtered album node.  Hmmmm, lots to study !! :)

DaveTBlake commented:
Eeek, you quoted my _long_ text. 
>The thing about filters and part of the reason I wasn't sure that a filter would be the correct way to go, is that as I understand it, a filter only applies at the current view 'level'.

The pass-across-levels flaw in the current filter type node implementation only happens  when the rule doesn't belong to the item e.g. a genre or path rule (song properties) applied to "artists" or "albums". I am in a big redesign of the rule -> sql part that will fix that.  The originial work was never done with music in mind and the db schema needs are different. Meanwhile this flaw it does not impact boxed sets because the rule belongs to above (i.e. album is a boxed set) and so holds.

But I totally understand why you took a folder approach, I have a lot of miles under my belt and only now really getting to grips with dealing with this (and what Montellse started doing in 2013 but didn't finish). In fact looking at your work has inspired me :)

What I am suggesting is still to use a folder/path approach but does not need so many new CDirectoryNode descendants. I have something almost working, so if you hang on a bit I will post my hack of your work and you can see what I mean.

DaveTBlake commented:
Have a look at my branch and see what you can learn from it.  You can fetch it locally using something like
```
git fetch https://github.com/DaveTBlake/xbmc/ Add_boxedsets2
git checkout FETCH_HEAD
git checkout -b Add_boxedsets_Dave
```
I will rebase if you need (it was based on your branch at some point) and have some other prefered way of using git. I am still a git novice really, I can do what I need and that's it.

Even if we don't go that direction there are still things you can pick from it and use. Do ask questions too.

I have had some more ideas this morning about making the node to child node navigation flexible - what I did so far isn't right, and it is a major restriction to lift. I will be back to talk about it later.

the-black-eagle committed:
Use filter and GetsongsByWhere in GetBoxsetDiscSongs

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -9032,86 +9032,40 @@ bool CMusicDatabase::GetBoxsetDiscs(const std::string& strBaseDir, CFileItemList
+  musicUrl.FromString(buildUrl);
+  musicUrl.AddOption("disctitle", strDiscName.c_str());
+  GetFilter(musicUrl, filter, sorting);
+  GetSongsByWhere(musicUrl.ToString(), filter, items);
```
OK, but not what I meant. You don't need this method at all with its bespoke path parsing
Filter using idAlbum and discnumber not disc title which may not be unique.  See https://github.com/DaveTBlake/xbmc/blob/20b51ce9ddbbd7db2655771e73dd15717e21051b/xbmc/music/MusicDatabase.cpp#L11184-L11202
Where you call GetBoxsetDiscSongs is just getting songs with that filter which can already be on the item path

the-black-eagle replied:
I know disc title may not be unique.  Certainly in the case of auto-generated names it won't be, but the combination of idAlbum and disc title will be.  Maybe discnumber is indeed better though.  I've only quickly had a look at your code.  I've built it and tested it against my small sample set and it works fine, although you seem to have broken the disc name that I added to estuary.  No biggie though.

the-black-eagle replied:
Oh, on the subject of what is displayed, there _is_ a bit of an issue when we stack the albums and set pItem->SetLabel(strDiscSubtitle).  In estuary (haven't tested with confluence yet, but will do) only two views out of the four show the correct disc title.  Shift and WideList both show correctly, Wall and InfoWall both show the album title, not the disc title.  I don't yet know if that's a skin thing or a core thing but I thought it might be worth mentioning as whichever 'thing' it is, it needs rectifying.  Both your version and mine display the same so I'm leaning towards skin issue, but I need to dig around in the skin files to see if that really is the case.

EDIT - Yeah, it's a skin thing.  Confluence works perfectly in all views.  I'll see if I can fix it, but I'm certainly no skinner.

DaveTBlake replied:
Possibly both a skin and a core issue. I know that my node hack has an issue with setting item list name (where what gets displayed should come from), but the skin could be doing its own thing on top of that especially if it varies with view type.

DaveTBlake commented on xbmc/filesystem/MusicDatabaseDirectory/DirectoryNodeBoxsetDiscSongs.cpp:
```diff
@@ -0,0 +1,49 @@
+  CollectQueryParams(params);
+
+  bool bSuccess=musicdatabase.GetBoxsetDiscSongs(BuildPath(), items);
+
```
What is this method and node doing that is different from `CDirectoryNodeSong`?

the-black-eagle replied:
Ah well, I guess I could ask you what's the difference between CDirectoryNodeSong and CDirectoryAlbumCompilationSongs because I don't think that is needed either.  You shouldn't need the "compilations" flag when getting the songs as all the albums in items are already compilations and thus returning the songs for each album is enough.  I've tested that funnily enough and it works fine.

DaveTBlake replied:
Yes exactly, CDirectoryAlbumCompilationSongs is historic and superseded. 
Understandably you followed how compilations nodes had been done for boxed sets, but it was an outdated approach.

DaveTBlake reviewed (commented):
Take a look at how I got rid of the extra DirectoryNodes, `CDirectoryNodeBoxsetDiscSongs` in particular I'm sure is unnecessary.

You should find my suggestions for CMusicDatabase::GetBoxsetDiscs https://github.com/DaveTBlake/xbmc/blob/20b51ce9ddbbd7db2655771e73dd15717e21051b/xbmc/music/MusicDatabase.cpp#L8948 helpful too - getting discnumber and title.

Try to think more about a generic discs node, that may be filtered and showing just the discs for albums that are boxed sets. Discs always have a number and may have a subtitle (of course those for boxed set always do).

DaveTBlake commented:
Ideally I think we want to get to a place where there are just as many music node types as there are types of content e.g. genres, artists, albums, _discs_, songs, sources, years, roles. Then the ability to navigate flexibly between them which still allows for the historic encoding of where you are in that navigation in a path (which I'm currently looking at).

Now you may think "can't I just add some DirectoryNodes to implement boxed sets handling now, and you clean up later", and I would understand, but now you have attracted my attention to the matter so it seems a good time to consider the outdated nodes issue generally :)

DaveTBlake commented:
If you are interested for a disc (rather than album) view of the music library we can use SQL like this
```
SELECT albumview.*, song.iTrack >> 16 AS discnum, song.strDiscSubtitle 
FROM albumview JOIN song on song.idAlbum = albumview.idAlbum
GROUP BY song.idAlbum, discnum
```
adding WHERE clauses to limit to just mutiple disc albums, or just boxed sets etc.  No problem not having a disc table etc., although we could add a view to the db schema if it was useful.

the-black-eagle commented:
> Now you may think "can't I just add some DirectoryNodes to implement boxed sets handling now, and you clean up later", and I would understand, but now you have attracted my attention to the matter so it seems a good time to consider the outdated nodes issue generally :)

Nah, that would be unfair.  Besides, I'd rather do it in a way that will move Kodi forwards instead of a way that could cause issues or complications in the future.  If stuff isn't really needed, or there is a better way of implementing something then I think its better to remove the redundant stuff sooner rather than later.

A good example of that is you splitting on disc number rather than title, because that can be re-used to split up any multi-disc album.  And yes, being able to do that _is_ something that I think is a benefit.  Who really wants to scroll through ~100 tracks on a 5 disc compilation to get to the disc they actually want to play ? I think it's much more natural to either navigate into or click play on, the actual disc you want to listen to.

Estuary uses ListItem.Title in those two views.  If I change that to ListItem.Label, then discs display correctly, but albums have the title displayed twice.   I only had time for a quick look though, so it could just be that I picked the wrong label to change.  I should have some more time later to have a better look at it.  

Oh, while I remember, I need to re-visit the auto-title bit and make the string localized instead of hard-coded.  Think I can use that #427 which, according to upstream master, just says 'disc'.

DaveTBlake commented:
Great, glad you are up for it :)
Meanwhile I think I have cracked having optional album > **discs** > songs navigation. This means rather than add 3 CDirectoryNode for handling boxed sets we just add one for discs generically. There is still  "Boxed sets" node (i.e. boxedset.xml that could be folder or filter type with content "albums") but not a CDirectoryNode or node type.
I have it on my repo in a new branch https://github.com/DaveTBlake/xbmc/tree/AddDiscsNode based on your commits, but see
https://github.com/DaveTBlake/xbmc/commit/c3572b7b8cc9119cb3de576fcbfca5256cd5338c
for the meat.

There are still things to tidy-up e.g.  add "*All discs", picking up folder art,  maybe a "discs" node (hence musicdb://discs etc.), and not sure about having a "discs" content for skinners. Look for "! @todo". Oh and `CDirectoryNodeBoxsetDiscs` and `GetBoxsetDiscs()` need renaming since they are no longer limited to boxed sets - `CDirectoryNodeDiscs` , and GetDiscsNav() perhaps? I left them as they were so you could see how it connected to your work so far.

I think this can be incorporated into your work. Since it removes things you added it could be easier to start a new branch, and pick the changes from this PR and my commit into it, it all depends how good you get at Git splitting and rebasing etc. to get a nice clear set of discrete commits (rather than edits on top of edits).

Anyway first thing is to try it, see what navigation from the albums node does depending on how many discs there are in the album.

the-black-eagle commented:
Me, good with git ?  Hahaha.  Google is my best friend in these cases though.  You are certainly a star.  I was just contemplating navigation like that as Estuary shows recently played/added albums on the right of the menu but obviously, clicking on one takes you to the full thing currently and I was thinking it would be much cooler if it took you to the individual discs instead.  I ran a test against my actual db earlier.  Exported it from my sql server and imported it locally, then adjusted advancedsettings.xml to point to the local machine.  Fired up Kodi and it correctly updated from 72 to 73, prompted me to re-scan the tags and all _almost_ looks good.  The only thing I can really pick up on is that I have a compilation album that has the boxset tag set and another copy (in a different source) that hasn't.  For some reason, one of the tracks doesn't update correctly in the db and I end up with one track that doesn't have strDiscSubtitle set for it. Originally, I hadn't forseen that and so I was ending up with 3 discs for a 2 disc set.  Same idAlbum but the number of rows returned from the db is 3, two titles and a blank line.  I avoided this by adding "And strDiscSubtitle != '' " to the query. [https://github.com/the-black-eagle/xbmc/blob/f1ab740644bec59d87e1e4aad8cc9b889395f122/xbmc/music/MusicDatabase.cpp#L8970](url).

Also with your branch, suddenly after the upgrade I have no disc titles for my Queen boxset, or any of the Erasure ones but all the rest **do** have titles.  That's a bit odd because I would have thought that if I'd mucked up something in Estuary that they all would be missing.  Not had chance yet to test against my own to see if its the upgrade responsible for it, or something else.  I'll get onto all this stuff as soon as I can.  Might look into a GUI for git as that might make things simpler.  It's all commandline stuff at the moment.

DaveTBlake commented:
Git advice....
On Windows I use a mix of cmd line and TortoiseGit as GUI. I think Bash is popular with the Linux guys. I know enough to get by, Google when I don't and ask the team when I mess up bigtime (not done that in a while, phew).

The final PR (yes we will get there!) will need to have at least separate commits for main functionality and JSON API, probably skinning stuff separate too. Each step needs to compile. Only touch those things where you actualy change functionality, cosmetic changes to other code parts need to be a separate PR (team happy to have code generally cleaned-up, but not mixed with features and fixes).

Obviously while working on something you create a commit trail that re-edits changes and isn't always coherent. I have often just made a copy branch (keeping the old WIP in case there is a bright idea I need to recover), and then squashed the lot in the new branch, and then rebased splitting into function areas if possible.

DaveTBlake commented:
As for missing strDiscSutitle values, since it is denormalised data we need to ensure that when populating the library all the songs on a disc have the same strDiscSubtitle value. The tagging may be inconsistent, but the library needs to consistent.

I have yet to pull your tag processing/library populating design to pieces :)
Got over excited about the GUI/Node side of things, but nothing will escape scrutny.

There is also an argument for storing a total discs count in the album table. Maybe not as it is easy enough to derive when needed, or could be added to albumview as a subquery (but slows things in MySQL, so I have it in mind to remove the subquery fields we have in fields as an optimisation - if we decide to care about MySQL that is). Also don't want it to be modified by NFO.
Denormalise, or not denormalise... always a question.

DaveTBlake commented:
Fixed the *all albums and added *all discs, and force pushed https://github.com/DaveTBlake/xbmc/commit/f59f236f872a55e1293c213c6b5bc1cb796d35e2 hope you had not already grabbed it and started picking/squishing etc.. If so the fetch it again.

the-black-eagle commented:
In backwards order :)
 No, I hadn't started on picking/squishing etc so that's fine.

Missing strDiscSubtitle. I have found what caused it !!  I dumped the tags for both copies of the album and found that for one track the mbtrack_id is different and the artist name and artistsort also differs slightly (although the artist_mbid is correct).  All the other songs have identical tags in each copy.  This album does not have any disc titles and was forced into the boxsets by adding the tag.  If you go the other way and add the album in the source that doesn't have the boxset tag, then the album is indeed removed from boxsets but that track shows up twice in the song list and in the db, strDiscSubtitle still contains "Disc 1" for one of them.  So that is a tagging issue, not a code issue. It would have happened irrespective of boxsets or discs.

Bash - yeah that's command line for linux (generally, there are other shells you can use).  I have a GUI (SmartGit) that lists commits and shows the changes between branches etc and apparently it's pretty powerful, I just have to learn how to use it !  In the meantime, I know a bit of commandline stuff, enough to be able to squash & fixup my commits but I am still on a huge learning curve, both with Kodi itself and git in general.  I'll get there no doubt though.

JSON - Bah, I had forgotten about that !!!  I'll get a clean branch with my stuff in and then I can mess about with json on this one.

DaveTBlake commented:
I would drop the JSON attempt, (clean it out the way or not more to the best branch) and sort out the rest first, including import/export of album bBoxset, art, smartplaylist rules (other core things you have not looked at. Only then get into JSON :)

For art Kodi will already pick up local art (folder.jpg thumb) per CD subfolder (I posted about it on the foum somewhere), but not sure if more is needed to have it show on the discs node. You will probably have better test data to play with that than I do. 

Plenty for you to do.  :)

the-black-eagle commented:
Cleaner branch with 3 commits -> https://github.com/the-black-eagle/xbmc/pull/2

Import/export was already working, at least in the tests I have done.  I think the most important aspect of art for this is discart.  Although individual discs may well come in paper sleeves or whatever in a set, the discs will definitely be different.  In fact, they generally are for any multi-disc album so I need to be looking at that.  Also, and you might need to ask a skinner about this, but I can't see a sensible way to alter ListItem.Title to be a disc title rather than the album title.  In some views in Estuary, ListItem.Label is displayed instead but in Wall and InfoWall views, title is displayed which means all the discs have the same original album name.

DaveTBlake commented:
Great you have managed the git manouvres :)
Won't have time to check it over until the weekend, but will continue comments/discussion on that PR when I get to it. 

Estuary issues we can get help with once the rest is done, possibly a separate PR for that. It is good to have a modified skin to see that the core changes and new info labels work etc., but generally skin changes themselves get separate PR. But make notes!

the-black-eagle commented:
Just a little thing I have noticed while adding art to my sets.    When you changed GetBoxsetDiscs, you added a line that sets the disc title to 'disc' if one doesn't exist however, one of the criteria for a boxset is that all the discs are titled, so that check is redundant for a boxset.  Also, splitting on disc number rather than strDiscSubtitle is fine for ordinary albums but I feel that it should still be title for boxsets as each title will be unique within that album.

This is because, for example, A tree full of secrets is a 9 volume set where each volume is two discs.  Originally, this is how my code split it up and each volume did indeed contain two discs.  Now it gets split into 18 different discs because we can't tell if a disc is part of a volume.  There will probably be other large sets that are also split up like this as it's a logical way to do it.

Just thought I'd better mention it as I've noticed it now.

DaveTBlake commented:
OK, now you bring up an entirely new thing, but better now than later :)
DISCSUBTITLE is not unique across the discs of a release -let's call all discs with the same non-null title a volume - and you want to split boxed sets by volume.

I am happy with adding "discs" as a new optional navigation level between albums and songs, that is useful in a variety of situations.
I am happy with flagging some albums as "boxed sets" i.e. complicated releases you possibly want to see separately. The previous new "discs" feature enables users expand those into discs.

But now some boxed sets may have another arbitrary grouping of songs. Not comfortable with this. Is it albums > volumes > discs > songs?? Saying group by disc title and only by disc when there isn't one gets messey when titles could be very inconsistent. Also disc number makes a nice UNIQUE id that works with node navigation path approach. Would need to have a "voume ID"...

Going to sleep on it and see if a nice idea comes, but really do we have to go that far? In those cases like "A Tree full of Secrets" can't the user tag volume as disc and ignore the original physical media? Maybe at that point you are talking about having sets of albums like sets of videos?

the-black-eagle commented:
Why aren't disc titles unique ?  Generally they either refer to an original album, a remaster of an album, a l live version or some sort of bonus disc.  I'm not denying that they may well not be unique over a set of albums, in that several sets could well have a 'bonus disc' in them, but in the concept of a boxset I have yet to see two discs with the same title (unless they are grouped by volume).  I have an idea about how this could work, but need a bit more time to see if it's really viable.  I don't want to break whats already implemented as it works really well, but i _would_ like to be able to do the volume thing if it doesn't affect anything else.
I often sleep on ideas and find that very often, I wake up with the answer in my head, so I await your thoughts with interest.

DaveTBlake commented:
>Why aren't disc titles unique ? ... in the concept of a boxset I have yet to see two discs with the same title (unless they are grouped by volume). 

That, different disc numbers with same subtitle.

No bright ideas came over night, so we take the other approach let it stew, discuss it back and forth :)
Splitting by disc is so nice and generic, splitting by disc or by subtitle sometimes is not. Why not split by other song properties - rating, playcount, lastplayed? I know in the context of a boxed set it may be meaningful, but there is still something arbitary yet too specific about it.

Multiple disc is going to be more common than boxed set, so really don't want to compromise that for those few with boxed sets in volumes. It would be great if Kodi UI engine made grouping/splitting things easy, but is was designed to have a fixed navigation structure instead.

the-black-eagle commented:
Multiple disc is indeed more common than boxed sets with volumes.  However, I don't see why we can't have the best of both worlds.  In spite of a little inflexibility in the UI engine, it's perfectly possible to do both things because of the 'rules' of a boxset.

We decided a boxset has to be three or more discs and that all those discs have to have subtitles in the tags. If an album doesn't have the subtitle tags _but_ a user thinks it should be a set, they can set the boxset tag which automatically generates a title for each disc that is missing one.  Therefore, there are titles for any disc in a boxset, as opposed to most other albums which might only have a title for one disc (bonus disc, live version, original version etc etc).
Taking that further, if an album has disc titles then it's likely (but not always necessarily true) that it's a boxset.  Obviously if there are _no_ titles, it definitely isn't a set and splitting on disc number is ideal.  If however there are titles then we can split on those.  This keeps volumes together but splits up everything else.  The only time I can see where there might possibly be an issue is if the first disc has a subtitle but the others do not.  However, upon thinking about it, it wouldn't be an issue because your code generates a 'disc %i' title for any disc without one.
Anyway, I had a bash at implementing it!  Turned out to not be too difficult really.  Keeps discs with the same title grouped together, splits up multi-disc albums into individual discs.  My Tree full of Secrets is now correctly showing two discs per volume and 9 volumes + the last disc (Volume 9 is only one disc).  2, 3, 4 and 5 CD sets that aren't boxsets also split correctly and show the right amount of discs as do the 'real' boxsets.  Currently I've only tested with a small subset of around 60 albums but I'll be expanding that later.  The test albums have one or two with disc names in only some of the discs, a couple with the boxset tag set, two that are mis-tagged on purpose and a bunch of 2, 3, 4 & 5 CD sets as well as some single CD albums (some with disc titles and some without).

DaveTBlake commented:
It is not fetching the data that concerns me, that is quite doable, it is what is the concept this creates, what is the node content? Not "discs", despite being based on DISCSUBTITLE, because splitting into discs gives a different result. Does the "volume"  then get split into actual discs? Calling it a volume just because the example you gave happend to use that in the DISCSUBTITLE tag isn't reasoned design either (just useful shothand for this discusson). It is arbitrary parts that might have meaning for a specific boxed set, or could be a tagging error. Even MB does n't list  9 volumes, it breaks the release _like all the other releases_ into 18 discs.  Adding an extra level of discs (based on disc number) to albums > songs is fine, taking DISCSUBTITLE as a well defined and well used tag that Kodi can base a node upon is not. Adding more than one new level to navigation with the intermediate level not having a generic definition and only aviallable is special circumstances is just not good. 

Imagine explaining it to certain users. Nagivation genres > artists >albums > parts/volumes (sometimes) > discs (if more than one and option enabled) >songs

the-black-eagle commented:
No, the 'volume' does not get split further.  If you open a volume you get the contents of both discs in that volume.  From a user point of view, that's what I'd expect to happen.  So it goes albums > discs/volumes > tracks.  I can happily agree that it's a niche area, but do users really want to look at all 18 discs at once or 9 volumes ?  For me, the latter is the better option and I've got no problem with opening a volume and seeing the contents of two discs inside, because that's whats meant to be in a volume.  In all other cases, the albums are split into individual discs.  That includes all other albums that have more than 1 disc.
I guess my reasons for wanting to do it this way is that it's more aesthetically pleasing in the GUI whereas your reasons against doing it are more technical.  There is no way I'd want to split up a very large set, say 30+ CD's into individual discs and navigate through those if I don't have to.  Even if it meant writing my own DISCSUBTITLE tags to sub-divide it into 2 or 3 discs at a time, I'd do that if I knew that Kodi could replicate that division on screen.  It's not another level, it's a way of telling Kodi that these x amount of discs go together in the same way that an album title says these tracks go together.  The navigation is still album > discs > tracks except that 'discs' has the possibility of containing the contents of more than one actual disc where grouping them together makes sense, either because MB have tagged them like that, or the user has.  In all the testing that I have done, this doesn't affect anything else apart from grouping together discs with the same title, which is presumably _why_ they are tagged with the same title in the first place.  Which is better to see on screen, Vol 1, Vol 1, Vol 2, Vol 2, Vol 3 etc or Vol 1, Vol 2, Vol 3 ?  If you were coming to Kodi for the first time, would you expect to see those duplicates, or expect it to be able to group them together ?

DaveTBlake commented:
>I guess my reasons for wanting to do it this way is that it's more aesthetically pleasing in the GUI whereas your reasons against doing it are more technical

Yes, that is exactly right. I do understand why as a user you want what I see as a fuzzy split into discs/volumes, but OTOH it is so niche conceptually. Setting out knowing the sometimes "discs" are not actually going to be discs just seems icky. "Disc groups"? No, not seriously. Also what about art for volumes, grouped discs, whatever they are? Having folders per disc, using disc number are nice generic ideas, you can't use title in the same way.

But I don't want to piss you off with a "no, no, no" attitude, so show me. I was half expecting an example of what you want to do to appear as a commit on #2. Perhaps do that. 

You also need to be aware that looking at CDirectoryNodes has resulted in https://github.com/DaveTBlake/xbmc/tree/NodeFolderLabelAll and a PR to Kodi which I will merge sooner or later but hope to get a review if I can find an interested team member. It could be worth pulling that in and rebasing on top of it.

DaveTBlake commented:
An approach, that I don't think you will like but is more generic, would be to spot where the same  non-null DISCSUBTITLE value has been set for more than one disc in the album (boxed set or not), and to append the disc number. So for that dreaded example you would see "Vol1 (Disc 1)", "Vol1 (Disc 2)", "Vol2 (Disc 3)", "Vol2 (Disc 4)" etc. , better than repeated titles on screen.

That would make me happy technically, the node type/level would always be "discs", not sometimes some other parts.  Also each volume of "A tree full of secrets" is pretty huge, surely it is more useful split into discs, and when you want to play a volume you add both separately?

the-black-eagle commented:
Backwards again !!!

Yeah it is pretty big, but 2 discs of tracks is a useable, manageable chunk.  There is a Rachmaninoff set on MB with 32 discs.  I'd love to be able to split that into 11 sets rather than 32 individual discs - it has no titles anyway so left like that it _would_ show as 32 discs.

Vol 1(disc 1) etc is better than just Vol 1, Vol 1.  

Art.  Kodi shows the right cover for each volume because each _disc_ has to have a cover otherwise one disc will show blank art when you play it.  So I still need to assign 18 covers although Kodi finds these and does it automatically.  If not, when scrolling down the list of tracks, when you get to the second disc, the art would disappear......I just know you're probably not going to like that although it's no different to assigning the art if you split it into individual discs.

Oh, I have a commit ready to go.  A couple actually.  One that adds the art stuff, one that does this and one that fixes a bug that I found today where the song table was not being updated correctly for an album that already existed in a different path.

Don't worry about pissing me off, it's extremely unlikely to happen no matter which way things eventually go.  Discussion is always good, as are other peoples points of view.  'Fuzzy' it might be, but fuzzy logic gives scope for flexibility where hard and fast rules do not.  It's probably not going to be seen by that many people, but the ones that _do_ see it will just think Kodi is smart.  If it doesn't go in, how long before someone shows up on the forum saying that they have discs with the same titles, why doesn't kodi group them ?  In fact, maybe some more user feedback about this would be helpful.  What do actual users want ?  The ones that want to be able to see proper sets, how do they want them split up ?

DaveTBlake commented:
>Backwards again !!!

Which bit, all of it?

>  In fact, maybe some more user feedback about this would be helpful. What do actual users want?

That is always tricky. The most vocal on the forum are not necessarily the most representative in opinion or use, but where else can you go. Then you have us team members, and we are always sure we are right even with features we don't use ;)  And of course the answers you get always depend on the way you ask the question.

Can we hammer away on this a bit more before raising users hopes/expectations with a forum posting. I want to look at what you have done, see if I can pick serious holes first. 

Art: you may have mixed up what is done by core and what is at the discretion of the skin, yet to check more carefully.

Great you are robust, I may be criticising your design/implementation but I am not attacking you. I am delighted that you have taken up this challenge.
